docs/_posts/2022-05-06-writing-cli-tools-with-typescript.md

```diff
@@ -0,0 +1,187 @@
+
+```typescript
+// src/yell.ts
+import { Command } from "commander";
+
+export default (): Command =>
+ new Command()
+  .name("yell")
+  .description("Say the word passed as the first argument but louder")
+  .argument("<word>")
+  .action((word: string) => console.log(word.toLocaleUpperCase()));
+```
+
+
+
+We also have to register our yell command with the main CLI command in src/cmd.ts, i.e. add an import statement for our subcommand and register it using *Commander.addCommand* 
+
+```typescript
+// src/cmd.ts
+// … 
+import buildYellCmd from "./yell";
+// …
+export default (): Command => {
+ const command = new Command()
+  .option("-g, --greet", `Say ${HELLO}`, false)
+  .addCommand(buildSayCmd())
+  .addCommand(buildYellCmd())
+  .addHelpCommand()
+  .showHelpAfterError();
+```
+
+
+
+```
+```
+
+Having done that, we can run a quick test on our command using the dev script in package.json. The dev script simply executes our TypeScript code with [ts-node](https://github.com/TypeStrong/ts-node).
+
+```bash
+$ npm run dev -- yell 'hey!'
+
+> @fkurz/typescript-cliutil-starter@1.0.0 dev /private/tmp/my-cli
+> ts-node src/main.ts "yell" "hey!"
+
+HEY!
+```
+
+We can also add a quick test by adding *src/yell.test.ts* with the following content
+
+```typescript
+import buildCmd from "./cmd";
+
+beforeEach(() => {
+ jest.spyOn(process, "exit").mockImplementation();
+ jest.spyOn(console, "log").mockImplementation();
+});
+
+describe("Yell subcommand", () => {
+ it("Should say 'HEY!' when passed 'hey!'", () => {
+  const cmd = buildCmd();
+  cmd.parse(["node", "cmd", "yell", "hey!"]);
+
+  expect(console.log).toHaveBeenCalledWith("HEY!");
+ });
+});
+```
+
+Since our Jest configuration expects JavaScript test files, we first have to transpile our code with the build script using the TypeScript compiler *tsc*. The transpiled JavaScript source code is put into the dist/ folder.
+
+```bash
+npm run build
+> @fkurz/typescript-cliutil-starter@1.0.0 build /private/tmp/my-cli
+> tsc --build
+```
+
+
+
+We can now run our test suite with Jest using the test script. New test files will automatically get recognized by Jest if they have—as in our case—the suffix *.test.js*
+
+```bash
+$ npm run test
+> @fkurz/typescript-cliutil-starter@1.0.0 test /private/tmp/my-cli
+> jest
+
+...
+
+Test Suites: 3 passed, 3 total
+Tests:       8 passed, 8 total
+Snapshots:   0 total
+Time:        0.41 s, estimated 1 s
+Ran all test suites.
+```
+As a final step, we can build a native binary by executing npm run package. The pkg tool will then bundle all source code and dependencies into a single binary file *bin/main*. 
+
+```bash
+$ npm run package
+
+> @fkurz/typescript-cliutil-starter@1.0.0 package /private/tmp/my-cli
+> pkg dist/main.js --no-bytecode --public-packages '*' --public --target host --output bin/main
+
+> pkg@5.3.1
+```
+
+Before running the program, we have to make it executable 
+
+```bash
+chmod +x ./bin/main
```

Comment:
Sounds like something that could be done in the `postpackage` script.

Reply:
Wouldn't that imply that we create a distribution archive (which we don't)?

> postpack

    Runs AFTER the tarball has been generated but before it is moved to its final destination (if at all, publish does not save the tarball locally)

(https://docs.npmjs.com/cli/v8/using-npm/scripts).

Reply:
Not `postpack` but `postpackage`.
Any `post<x>` script runs right after `<x>`, while `pre<x>` will be executed before `<x>`.

So if you run your `package` script
* the `prepackage` script (if it exists) will be executed before that while
*the `postpackage` script (again: if it exists) will be executed afterwards.

This can be done for any custom script.

https://docs.npmjs.com/cli/v8/using-npm/scripts#pre--post-scripts

Reply:
Ah ok, wasn't aware of that.

Reply:
I added a `postpackage` script in the repo.